src/node/rpc/member_frontend.h

```diff
@@ -1231,6 +1248,18 @@ namespace ccf
         {
           record_cose_governance_history(
             ctx.tx, member_id.value(), cose_auth_id->envelope);
+          if (is_proposal_replayed_or_stale(
+                ctx.tx,
+                cose_auth_id->protected_header.gov_msg_created_at,
+                request_digest,
+                proposal_id))
+          {
+            ctx.rpc_ctx->set_error(
+              HTTP_STATUS_BAD_REQUEST,
+              ccf::errors::DuplicateOrStaleProposal,
+              "Duplicate or stale proposal");
```

Comment:
I think we can distinguish these cases (duplicate proposal vs stale proposal), so we should do so in the error message? For a duplicate proposal, we can return the conflicting proposal ID (so if 2 idempotent-y tools try to submit the same proposal, they can both defer to talking about the state of the active instance). For a stale proposal, we can tell them how stale they are, and hint to freshen their `gov_msg_created_at` header.

Reply:
We could do something like that, but:

1. We have a bar for stale, but it doesn't mean the proposal isn't a duplicate. It's just that we have stopped keeping track of proposals in that time range.
2. The bar for stale isn't fixed, because it's based on the median of the timestamps of the proposals submitted successfully so far. So we can return the value we used to make the decision, but there's no guarantee that if you re-sign later than it, the proposal will be accepted.
3. We can return the proposal ID, but it feels like a half-step. The current interface doesn't try to be idempotent, the more (most?) idempotent version would return success (and the proposal id) on duplicate, and only an error when the proposal is older than the stale bar. It's still not fully idempotent, because making the same call twice only works in a limited time window.

I think one limitation with going with 3. (or indeed returning the proposal ID in a failure), is that we will delete the proposal content from the KV much sooner (*) than it will slide out of the replay detection window. And so there is typically little chance that the user can query via an online API for that proposal ID if it has already concluded (passed or rejected, and since deleted).

(*) on the next proposal.

Reply:
As discussed, going with errors in both cases, and detailed error messages.